File: segundoPeriodo/AEDS-2/estudos/tadsFlexiveis/fila/c/main.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Celula {
    int elemento;
    struct Celula *prox;
} Celula;

typedef struct Fila {
    Celula *primeiro;
    Celula *ultimo;
} Fila;

Celula *criarCelula(int x) {
    Celula *nova = (Celula*)malloc(sizeof(Celula));
    nova->elemento = x;
    nova->prox = NULL;
    return nova;
}

Fila *criarFila() {
    Fila *fila = (Fila*)malloc(sizeof(Fila));
    fila->primeiro = criarCelula(0);
    fila->ultimo = fila->primeiro;
    return fila;
}

void inserir(Fila *fila, int x) {
    Celula *nova = criarCelula(x);
    fila->ultimo->prox = nova;
    fila->ultimo = nova;
    nova = NULL;
}

int remover(Fila *fila) {
    if(fila->primeiro == fila->ultimo) {
        return -1;
    }
    Celula *tmp = fila->primeiro->prox;
    int resp = tmp->elemento;
    fila->primeiro->prox = tmp->prox;
    free(tmp);
    tmp = NULL;
    return resp;
}

void mostrar(Fila *fila) {
    for(Celula *i = fila->primeiro->prox; i != NULL; i = i->prox) {
        printf("%d ", i->elemento);
    }
    printf("\n");
}

void freeFila(Fila *fila) {
    Celula *i = fila->primeiro;
    while(i != NULL) {
        Celula *tmp = i->prox;
        free(i);
        i = tmp;
    }
    free(fila);
}

int main() {
    Fila *fila = criarFila();
    inserir(fila, 3);
    inserir(fila, 5);
    inserir(fila, 7);
    mostrar(fila); // 3 5 7
    printf("Elemento %d removido\n", remover(fila));
    mostrar(fila); // 5 7
    freeFila(fila);
    return 0;
}

File: segundoPeriodo/AEDS-2/estudos/tadsFlexiveis/fila/java/Main.java
class Celula {
    public int elemento;
    public Celula prox;

    public Celula() {
        this.elemento = 0;
        this.prox = null;
    }
    public Celula(int x) {
        this.elemento = x;
        this.prox = null;
    }
}

class Fila {
    private Celula primeiro;
    private Celula ultimo;
    
    public Fila() {
        primeiro = new Celula();
        ultimo = primeiro;
    }

    public void inserir(int x) {
        Celula nova = new Celula(x);
        ultimo.prox = nova;
        ultimo = nova;
        nova = null;
    }

    public int remover() {
        if(primeiro == ultimo) {
            throw new IllegalArgumentException("Fila vazia");
        }
        Celula tmp = primeiro.prox;
        int resp = tmp.elemento;
        primeiro.prox = tmp.prox;
        tmp.prox = null;
        tmp = null;
        return resp;
    }

    public void mostrar() {
        for(Celula i = primeiro.prox; i != null; i = i.prox) {
            System.out.printf("%d ", i.elemento);
        }
        System.out.printf("\n");
    }
}

public class Main {
    public static void main(String args[]) {
        Fila fila = new Fila();
        fila.inserir(3);
        fila.inserir(5);
        fila.inserir(7);
        fila.mostrar(); // 3 5 7
        System.out.println("Elemento " + fila.remover() + " removido");
        fila.mostrar(); // 5 7
    }
}